### ISSUE TITLE
AVD Alerts Bug: Issue with API Limitations

### ISSUE THREAD

**Issue by brmcmill** [Azure/avdaccelerator#392]
Ran into this error deploying to a sub in Azure Gov.  May be due to limitations with APIs.

The script that timed out: https://github.com/Azure/avdaccelerator/blob/main/workload/scripts/alerts/dsHostPoolVMMap.ps1

Referenced an article below for help.

[Troubleshooting throttling errors in Azure - Virtual Machines | Microsoft Learn](https://learn.microsoft.com/en-us/troubleshoot/azure/virtual-machines/troubleshooting-throttling-errors)

```JSON
{
    "status": "Failed",
    "error": {
        "code": "DeploymentFailed",
        "target": "/subscriptions/.../providers/Microsoft.Resources/deployments/carml_ds-PS-GetHostPoolVMAssociation",
        "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see [https://aka.ms/arm-deployment-operations](https://aka.ms/arm-deployment-operations) for usage details.", 
            {
                "code": "ResourceDeploymentFailure",
                "target": "/subscriptions/.../providers/Microsoft.Resources/deploymentScripts/ds_GetHostPoolVMAssociation",
                "message": "The resource write operation failed to complete successfully, because it reached terminal provisioning state 'failed'.",
                "details": [
                    {
                        "code": "DeploymentScriptError",
                        "message": "The provided script failed with the following error:\r\nNewtonsoft.Json.JsonSerializationException: Error converting value \"{\r\n  \"error\": {\r\n    \"details\": [\r\n      {\r\n        \"code\": \"TooManyRequests\",\r\n        \"message\": \"{\\\"operationGroup\\\":\\\"HighCostGet3Min\\\",\\\"startTime\\\":\\\"2023-06-16T15:22:24.0621174+00:00\\\",\\\"endTime\\\":\\\"2023-06-16T15:23:30+00:00\\\",\\\"allowedRequestCount\\\":140,\\\"measuredRequestCount\\\":190}\",\r\n        \"target\": \"HighCostGet3Min\"\r\n      }\r\n    ],\r\n    \"innererror\": {\r\n      \"internalErrorCode\": \"TooManyRequestsReceived\"\r\n    },\r\n    \"code\": \"OperationNotAllowed\",\r\n    \"message\": \"The server rejected the request because too many requests have been received for this subscription.\"\r\n  }\r\n}\" to type 'Microsoft.Azure.Commands.Compute.Models.PSComputeLongRunningOperation'. Path '', line 1, position 686.\n ---> System.ArgumentException: Could not cast or convert from System.String to Microsoft.Azure.Commands.Compute.Models.PSComputeLongRunningOperation.\n   at Newtonsoft.Json.Utilities.ConvertUtils.EnsureTypeAssignable(Object value, Type initialType, Type targetType)\n   at Newtonsoft.Json.Utilities.ConvertUtils.ConvertOrCast(Object initialValue, CultureInfo culture, Type targetType)\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)\n   --- End of inner exception stack trace ---\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)\n   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)\n   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)\n   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)\n   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value)\n   at Microsoft.Azure.Commands.Compute.Common.ComputeCloudException.GetErrorMessageWithRequestIdInfo(CloudException cloudException)\n   at Microsoft.Azure.Commands.Compute.Common.ComputeCloudException..ctor(CloudException ex)\n   at Microsoft.Azure.Commands.Compute.ComputeClientBaseCmdlet.ExecuteClientAction(Action action)\n   at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()\r\nat <ScriptBlock>, /mnt/azscripts/azscriptinput/dsHostPoolVMMap.ps1: line 43\r\nat <ScriptBlock>, <No file>: line 1\r\nat <ScriptBlock>, /mnt/azscripts/azscriptinput/DeploymentScript.ps1: line 295. Please refer to [https://aka.ms/DeploymentScriptsTroubleshoot](https://aka.ms/DeploymentScriptsTroubleshoot) for more deployment script information."
                    }
                ]
            }
        ]
    }
}
```

**Comment by JCoreMS** [comment]
This should be fixed but will wait to verify.  Added some delay in the deployment script to hopefully offset the API calls and not hit the limit.
start-sleep commands now between host Pool and Session host loops:
https://github.com/Azure/avdaccelerator/blob/main/workload/scripts/alerts/dsHostPoolVMMap.ps1

**Comment by JCoreMS** [comment]
Still in progress.... the added delay did not address the issue. This appears to only impact large environments with 50 or more VMs in a single Host Pool.  Researching alternate ways to map the actual VMs to the SessionHosts for AVD.

**Comment by JCoreMS** [comment]
Fixed with v2.1.0 - no longer a deployment script in the solution which should resolve this issue.  @brmcmill thanks for testing and confirming.

**Comment by brmcmill** [comment]
Tested it, it works great!  Thanks again for the support @JCoreMS.  If you'd like to close, you have my blessing 👍
